examples/async_client_hello_callback/src/bin/quic_async_client_hello_callback_server.rs

```diff
@@ -0,0 +1,146 @@
+use tokio::{fs, sync::OnceCell};
+
+/// NOTE: this certificate is to be used for demonstration purposes only!
+pub static CERT_PEM_PATH: &str = concat!(
+    env!("CARGO_MANIFEST_DIR"),
+    "/../../quic/s2n-quic-core/certs/cert.pem"
+);
+/// NOTE: this certificate is to be used for demonstration purposes only!
+pub static KEY_PEM_PATH: &str = concat!(
+    env!("CARGO_MANIFEST_DIR"),
+    "/../../quic/s2n-quic-core/certs/key.pem"
+);
+
+type Sni = String;
+
+struct MyAsyncCertLoaderCHCallback {
+    cache: Cache<Sni, Arc<OnceCell<Config>>>,
+}
+
+impl MyAsyncCertLoaderCHCallback {
+    fn new() -> Self {
+        MyAsyncCertLoaderCHCallback {
+            // store Configs for up to 100 unique SNIs
+            cache: Cache::new(100),
+        }
+    }
+}
+
+impl ClientHelloCallback for MyAsyncCertLoaderCHCallback {
+    fn on_client_hello(
+        &self,
+        connection: &mut Connection,
+    ) -> Result<Option<Pin<Box<dyn ConnectionFuture>>>, S2nError> {
+        let sni = connection
+            .server_name()
+            .ok_or_else(|| S2nError::application(Box::new(CustomError)))?
+            .to_string();
+
+        let once_cell_config = self
+            .cache
+            .get_with(sni.clone(), || Arc::new(OnceCell::new()));
+        if let Some(config) = once_cell_config.get() {
+            connection.set_config(config.clone())?;
+            // return `None` if the Config is already in the cache
+            return Ok(None);
+        }
+
+        let fut = async move {
+            let fut = once_cell_config.get_or_try_init(|| async {
+                let fail = rand::random();
+                eprintln!("simulating network failure. failed: {}", fail);
+                if fail {
+                    return Err(S2nError::application(Box::new(CustomError)));
+                }
+
+                // load the cert and key file asynchronously.
+                let (cert, key) = {
+                    // the SNI can be used to load the appropriate cert file
+                    let _sni = sni;
+                    let cert = fs::read_to_string(CERT_PEM_PATH)
+                        .await
+                        .map_err(|_| S2nError::application(Box::new(CustomError)))?;
+                    let key = fs::read_to_string(KEY_PEM_PATH)
+                        .await
+                        .map_err(|_| S2nError::application(Box::new(CustomError)))?;
+                    (cert, key)
+                };
+
+                // sleep(async tokio task which doesn't block thread) to mimic delay
+                tokio::time::sleep(Duration::from_secs(5)).await;
+
+                let mut config = Config::builder();
+                config
+                    .load_pem(cert.as_bytes(), key.as_bytes())?
+                    .enable_quic()?
+                    .set_security_policy(&s2n_tls::security::DEFAULT_TLS13)?
+                    .set_application_protocol_preference([b"h3"])?;
+                config.build()
```

Comment:
I think we should encourage using the `Server::builder()` instead of a normal s2n-tls config, since they have to remember to enable things for it to actually work. We'd need to add a method to get the config out, though. Or `impl From<s2n_quic_tls::Server> for s2n_tls::Config {}`

```rust
Server::builder()
    .with_certificate(cert, key)?
    .build()?
    .into()
```